### DISCUSSION TITLE
How to make a footer with left and right aligned content?

### DISCUSSION THREAD

**Discussion by jaads** [dolanmiu/docx#940] [Q&A]
I want to have some text at the most left side of the footer and the page numbering at the most right, but both at the same height.
how can I accomplish this?

**Comment by dolanmiu** [comment]
Try using tab stops:

https://docx.js.org/#/usage/tab-stops

**Reply by jaads** [reply]
thanks. i just tried it but apparently it's not working. 
here is my code (I'm on version 5.5.0). I also tried the paragraph below outside of a footer object, but there also not tab is displayed. In both cases the result in the docx file is just e.g. "hellopage: 2"

```javascript
new Footer({
            children: [
                new Paragraph({
                    children: [
                        new TextRun({
                            text: "hello".toUpperCase(),
                            font: "Calibri",
                        }),
                        new TextRun({
                            children: ["page: ", PageNumber.CURRENT],
                            font: "Calibri",
                        }),
                    ],
                    tabStops: [
                        {
                            type: TabStopType.RIGHT,
                            position: TabStopPosition.MAX,
                        },
                    ],
                }),
            ],
        })
```

**Reply by aviau** [reply]
Its not working in your example because you don't have any tab.

You need something like this:

```js
                        new ImageRun({
                            data: headerLogo,
                            transformation: {
                                width: 50,
                                height: 50,
                            }
                        }),
                        new TextRun('\t'),
                        new ImageRun({
                            data: headerText,
                            transformation: {
                                width: 105,
                                height: 50,
                            },
                        }),
```

**Reply by aviau** [reply]
On my end however this only works in Word, not google docs.

I wish there was something that was well supported in both viewers.

**Comment by BobPritchett** [comment]
It looks like Word uses alignment tabs, which it doesn't appear the code generates. For example, 

```
<w:PTab Alignment="CENTER" RelativeTo="MARGIN" Leader="NONE"></w:PTab>
<w:PTab Alignment="RIGHT" RelativeTo="MARGIN" Leader="DOT"></w:PTab>
```

<img width="154" alt="image" src="https://user-images.githubusercontent.com/26860966/209019464-d4b50236-c324-4cdb-8139-b9d172b1b993.png">

In Word you can insert these into a Header/Footer with "Insert Alignment Tab". It would be nice to have support here, so that you can center things in the header or footer without knowing the margins to calculate a center tab stop.

**Reply by dolanmiu** [reply]
It is possible to do this:

https://github.com/dolanmiu/docx/blob/master/src/file/paragraph/formatting/tab-stop.ts#L52-L62

Usage:
https://github.com/dolanmiu/docx/blob/master/demo/75-tab-stops.ts

**Reply by BobPritchett** [reply]
I'm sorry, but I'm not seeing how... there doesn't seem to be a way to set the relativeTo attribute, and there is a distinction between tabs and PTabs that matters. The docx examples linked need a position. 

With PTab and relativeTo you can set a tab to the center of the margin area and it ignores other tab stop and goes directly there. (See https://learn.microsoft.com/en-us/dotnet/api/documentformat.openxml.wordprocessing.positionaltab?view=openxml-2.8.1)

**Reply by dolanmiu** [reply]
Thanks
I’ll investigate more tomorrow

I found some additional documentation on it too:

https://c-rex.net/projects/samples/ooxml/e1/Part4/OOXML_P4_DOCX_ptab_topic_ID0EKZDQ.html

**Reply by dolanmiu** [reply]
I have done the feature. It's currently sitting in a PR!

https://github.com/dolanmiu/docx/pull/1852

Works pretty well. I learnt more about Tabs and Positional tabs, added more missing `Run` elements also, along with the PR.

**Reply by Kaishley** [reply]
Hi Dolan, when will this new feature be released to npm?

**Reply by dolanmiu** [reply]
In version 8, after https://github.com/dolanmiu/docx/pull/1960 is completed.
